File: backend/app/services/booking_service.py
from datetime import datetime
from typing import List, Dict, Any, Optional
from fastapi import HTTPException, BackgroundTasks

from sqlalchemy.orm import Session

from ..models.booking import BookingCreate, Booking
from ..repositories.booking_repository import BookingRepository
from ..services.email_service import EmailService
from ..db.models import User, Slot

class BookingService:
    def __init__(self):
        self.repository = BookingRepository()
        self.email_service = EmailService()

    def create_booking(
        self, db: Session, booking_data: BookingCreate, background_tasks: BackgroundTasks
    ) -> Booking:
        """
        Create a new booking with duplicate prevention,
        and send a confirmation email in the background
        """
        # Create the booking in the database
        db_booking = self.repository.create_booking(db, booking_data)

        # Fetch additional data needed for email
        user = db.query(User).filter(User.id == booking_data.user_id).first()
        slot = db.query(Slot).filter(Slot.id == booking_data.slot_id).first()

        if user and slot:
            # Prepare data for email
            email_data = {
                "id": db_booking.id,
                "user_name": f"{user.first_name} {user.last_name}",
                "date": db_booking.booking_date.strftime("%A, %B %d, %Y"),
                "time": db_booking.booking_date.strftime("%I:%M %p"),
                "location": slot.location.name if hasattr(slot, 'location') else "N/A"
            }

            # Send confirmation email asynchronously
            self.email_service.send_booking_confirmation_async(
                background_tasks, user.email, email_data
            )

        # Convert DB model to Pydantic model for response
        return Booking.from_orm(db_booking)

    def get_user_bookings(self, db: Session, user_id: int) -> List[Booking]:
        """Get all bookings for a specific user"""
        db_bookings = self.repository.get_user_bookings(db, user_id)
        return [Booking.from_orm(booking) for booking in db_bookings]

    def get_booking_by_id(self, db: Session, booking_id: int) -> Optional[Booking]:
        """Get a booking by its ID"""
        db_booking = self.repository.get_booking_by_id(db, booking_id)
        if not db_booking:
            return None
        return Booking.from_orm(db_booking)

    def cancel_booking(self, db: Session, booking_id: int) -> Optional[Booking]:
        """Cancel a booking"""
        db_booking = self.repository.cancel_booking(db, booking_id)
        if not db_booking:
            return None
        return Booking.from_orm(db_booking)

File: backend/src/services/bookingService.ts
import bookingRepository from '../repositories/bookingRepository';
import emailService from './emailService';
import userRepository from '../repositories/userRepository';
import logger from '../utils/logger';

class BookingService {
  async createBooking(bookingData: any): Promise<any> {
    try {
      // Validate booking data
      if (!bookingData.userId || !bookingData.locationId || !bookingData.slotId || !bookingData.date) {
        throw new Error('Missing required booking information');
      }

      // Check for duplicate bookings
      const hasConflict = await bookingRepository.checkConflictingBooking(
        bookingData.locationId,
        bookingData.slotId,
        new Date(bookingData.date)
      );

      if (hasConflict) {
        logger.warn(`Duplicate booking attempt: slot already booked for location ${bookingData.locationId}, slot ${bookingData.slotId}, date ${bookingData.date}`);
        throw new Error('This slot is already booked for the selected date and location');
      }

      // Create the booking
      const booking = await bookingRepository.create({
        ...bookingData,
        status: 'pending'
      });

      // Get user for email notification
      const user = await userRepository.findById(bookingData.userId);
      if (!user) {
        logger.error(`User not found for booking: userId=${bookingData.userId}`);
        throw new Error('User not found');
      }

      // Send confirmation email
      await emailService.sendBookingConfirmation(booking, user);

      // Update booking status to confirmed
      const [, updatedBookings] = await bookingRepository.updateStatus(booking.id, 'confirmed');

      logger.info(`Booking created and confirmed: ${booking.id}`);
      return updatedBookings[0] || booking;
    } catch (error) {
      logger.error(`Booking creation failed: ${error}`);
      throw error;
    }
  }

  async getUserBookings(userId: number): Promise<any[]> {
    try {
      if (!userId) {
        throw new Error('User ID is required');
      }

      return await bookingRepository.findByUserId(userId);
    } catch (error) {
      logger.error(`Failed to get user bookings: ${error}`);
      throw error;
    }
  }

  async cancelBooking(bookingId: number, userId: number): Promise<any> {
    try {
      const booking = await bookingRepository.findById(bookingId);

      if (!booking) {
        throw new Error('Booking not found');
      }

      if (booking.userId !== userId) {
        logger.warn(`Unauthorized cancellation attempt: User ${userId} attempted to cancel booking ${bookingId} owned by user ${booking.userId}`);
        throw new Error('Unauthorized: You can only cancel your own bookings');
      }

      const [, updatedBookings] = await bookingRepository.updateStatus(bookingId, 'cancelled');
      logger.info(`Booking ${bookingId} cancelled by user ${userId}`);

      return updatedBookings[0];
    } catch (error) {
      logger.error(`Booking cancellation failed: ${error}`);
      throw error;
    }
  }
}

export default new BookingService();
